### PR TITLE
Pass pids to system probe for gathering IO metrics

### PR DESCRIPTION
**Pull request by wiyu**
### What does this PR do?

Pass PIDs to the system-probe to avoid having to iterate through `/proc/` 

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes
1. Enable the System probe and the process module in system probe
```
system_probe_config:
  enabled: true
  process_config:
    enabled: true
  log_level: debug
```

1. Validate the JSON endpoint is returning valid results from 
```
curl -s -H "Content-Type: application/json" -X POST localhost/proc/stats --unix-socket /opt/datadog-agent/run/sysprobe.sock -d '{"pids":[1]}' | jq .
{
  "statsByPID": {
    "1": {
      "openFDCount": 74,
      "readCount": "4552976",
      "writeCount": "2282360",
      "readBytes": "302642176",
      "writeBytes": "62115840"
    }
  }
}
```
- Verify `openFDCount` matches `/proc/fd/<pid>`
     `sudo ls /proc/1/fd | wc -l`
- Verify IO metrics look right. 
    ```
    $ sudo cat /proc/1/io
    rchar: 1292273055
    wchar: 370459539
    syscr: 4553018
    syscw: 2282360
    read_bytes: 302642176
    write_bytes: 62115840
    cancelled_write_bytes: 6455296
    ```
2. Verify Process-agent is calling the system probe (regular and real-time collection)
  - Check the `system-probe.log`
  ```
  2021-06-01 14:08:23 UTC | SYS-PROBE | INFO | (cmd/system-probe/modules/process.go:91 in logProcTracerRequests) | Got request on /proc/stats (count: 3): retrieved 1 stats in 215.571µs
```

### CODE REVIEW

File: pkg/process/net/common.go [hunk 1]
```diff
@@ -81,12 +83,23 @@ func GetRemoteSystemProbeUtil() (*RemoteSysProbeUtil, error) {
...
+	procReq := &model.ProcessRequest{
+		Pids: pids,
+	}
+
+	reqBody, err := reqEncoding.GetMarshaler(reqEncoding.ContentTypeProtobuf).Marshal(procReq)
 	if err != nil {
 		return nil, err
 	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s", procStatsURL), bytes.NewReader(reqBody))
```
Comment (shang-wang) [inline pkg/process/net/common.go:96]: nit: sorry this is probably the same with original code, but I think we don't need the `fmt.Sprintf` part.

File: pkg/process/encoding/request/encoding.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package request
+
+import (
+	"github.com/gogo/protobuf/jsonpb"
```
Comment (shang-wang) [inline pkg/process/encoding/request/encoding.go:4]: nit: I think this line should be separated from native imports.

**Review by shang-wang — APPROVED** [review]
Looks good, thanks

File: cmd/system-probe/modules/process.go [hunk 3]
```diff
@@ -49,11 +51,22 @@ func (t *process) GetStats() map[string]interface{} {
...
-		stats, err := t.probe.StatsWithPermByPID()
+		pids, err := getPids(req)
+		if err != nil {
+			log.Errorf("%s", err)
+			w.WriteHeader(http.StatusBadRequest)
+		}
+
+		stats, err := t.probe.StatsWithPermByPID(pids)
 		if err != nil {
 			log.Errorf("unable to retrieve stats using process_tracer: %s", err)
```
Comment (shang-wang) [inline cmd/system-probe/modules/process.go:68]: nit: Can we update this line? The module was originally called `process_tracer` but renamed to just process
```suggestion
			log.Errorf("unable to retrieve stats using process module: %s", err)
```

**Comment by brycekahle** [comment]
I have concerns about this approach from a security PoV. It would be an unprivileged process controlling a privileged process' behavior, and also parsing of data (the request body) supplied by an external entity.

File: cmd/system-probe/modules/process.go [hunk 4]
```diff
@@ -49,11 +51,22 @@ func (t *process) GetStats() map[string]interface{} {
...
+			return
+		}
+
 		start := time.Now()
-		stats, err := t.probe.StatsWithPermByPID()
+		pids, err := getPids(req)
 		if err != nil {
-			log.Errorf("unable to retrieve stats using process_tracer: %s", err)
-			w.WriteHeader(500)
+			log.Errorf("%s", err)
```
Comment (xornivore) [inline cmd/system-probe/modules/process.go:62]: Nit, you could use `log.Error()` instead of `log.Errorf`.

**Review by brycekahle — APPROVED** [review]

**Review by xornivore — APPROVED** [review]

**Review by sgnn7 — APPROVED** [review]

File: cmd/system-probe/modules/process.go (hunk 4 above)
Comment (wiyu) [reply cmd/system-probe/modules/process.go]: Fixed to add an actual message and not just the error.

**Pull request merged**